server/main.cpp

```diff
@@ -104,6 +112,165 @@ int create_listen_socket()
+    const char* member = "StartUnit";
+    const char* mode = "replace";
+
+    int ret = 0;
+
+    // Connect to the system bus
+    ret = sd_bus_open_system(&bus);
+    if (ret < 0) {
+        std::cerr << "Failed to connect to system bus: " << strerror(-ret) << std::endl;
+        return ret;
+    }
+
+    // Call the method to start the service
+    ret = sd_bus_call_method(bus, destination, path, interface, member, &error, &msg, "ss", service_name.c_str(), mode);
+    if (ret < 0) {
+        std::cerr << "Failed to start service: " << error.message << std::endl;
+        goto finish;
+    }
+
+finish:
+    sd_bus_error_free(&error);
+    sd_bus_message_unref(msg);
+    sd_bus_unref(bus);
+    return ret;
+}
+
+bool is_service_active(const std::string& service_name) {
+    sd_bus* bus = nullptr;
+    sd_bus_error error = SD_BUS_ERROR_NULL;
+    sd_bus_message* msg = nullptr;
+    const char* destination = "org.freedesktop.systemd1";
+    const char* path = "/org/freedesktop/systemd1";
+    const char* interface = "org.freedesktop.systemd1.Manager";
+    const char* member = "GetUnit";
+
+    int ret = 0;
+
+    // Connect to the system bus
+    ret = sd_bus_open_system(&bus);
+    if (ret < 0) {
+        std::cerr << "Failed to connect to system bus: " << strerror(-ret) << std::endl;
+        return false;
+    }
+
+    // Call the method
+    ret = sd_bus_call_method(bus, destination, path, interface, member, &error, &msg, "s", service_name.c_str());
+    if (ret < 0) {
+        std::cerr << "Failed to issue method call: " << error.message << std::endl;
+        goto finish;
+    }
+
+    // Parse the response message
+    const char* object_path;
+    ret = sd_bus_message_read(msg, "o", &object_path);
+    if (ret < 0) {
+        std::cerr << "Failed to parse response message: " << strerror(-ret) << std::endl;
+        goto finish;
+    }
+
+    // Check the service's ActiveState property
+    const char* active_state;
+    ret = sd_bus_get_property_string(bus, destination, object_path, "org.freedesktop.systemd1.Unit", "ActiveState", &error, &active_state);
+    if (ret < 0) {
+        std::cerr << "Failed to get ActiveState: " << error.message << std::endl;
+        goto finish;
+    }
+
+    bool is_active = (std::string(active_state) == "active");
+
+finish:
+    sd_bus_error_free(&error);
+    sd_bus_message_unref(msg);
+    sd_bus_unref(bus);
+    return is_active;
+}
+
+pid_t start_application_systemd()
+{
+	std::string service_name = "wivrn-application.service";
+    pid_t pid;
+    int ret = get_service_pid(service_name, pid);
+
+    if (ret < 0 || pid == 0) {
+        ret = start_service(service_name);
+        if (ret < 0) {
+            std::cerr << "Failed to start service " << service_name << std::endl;
+            return ret;
+        }
+
+        // Wait until the service is active
+        while (!is_service_active(service_name)) {
+            std::this_thread::sleep_for(std::chrono::seconds(0.1));
```

Comment:
`std::chrono::seconds` takes integer values, it should be written in milliseconds.